project/OscarPolai/Readme.md

```diff
@@ -0,0 +1,111 @@
+That it, knowing how dataframe do its memory allocation is important.  Also knowing what kind of operation underground will avoid memory waste.
+In this project, I will try to write the code and test the memory behavior parallelly to get more knowledge.
+
+Usually,  there are two ways to implement arrays of composite types. SoA or the columnar arrays perform
+better performance (cache locality) than AoS for heterogeneous data, especially when there are many field. 
+
+For this purpose, designing a data structure handing large datasets with high performance is the core of this project I am focusing. It means not only to implement the pure 
+Dataframe, but also design better data structures or operation for dataframe to reduce memory overhead.
+  
+Prospective Users
+=================   
+
+1. Those who do data analysis dealing with large datasets.
+2. Anyone who wants customize their library to enhance the flexibility and functionality pandas can’t do.
+
+
+System Architecture
+===================
+
+A main DataFrame class will be written in C++ and connect to python script with pybind11. 
+
+API Description
+===============
+
+Main class will implement some of the most common and useful.
+
+Here only list some of the most important method, if time allows, I will try to implement more.
+
+* Constructor to Load data 
+    * Dataframe(std::string csv_name);   // load data from csv file 
+* Load data from class method
+
+```cpp
+#include <DataFrame/DataFrame.h>
+
+DataFrame df;
+std::vector<unsigned long>  idx_col1 = { 1, 2, 3, 4, 5};
+std::vector<std::string>    col1 = { "A", "B", "C", "D", "E"};
+std::vector<std::string>    col2 = { "QQ", "Hello World!", "man", "M", "N"};
+
+df.load_data(std::move(idx_col1),
+                     std::make_pair("string col", col1),
+                     std::make_pair("Happy", col2));
+
+```
+* Popular methods
+    * Dataframe.get_column(std::col_name); // get column from specific column name
+    * Dataframe.drop_col(std::string col_name); //drop column by specifying name.
+    * Dataframe.cols(); // return the number of columns in dataframe
+    * Dataframe.rows(); //  return the number of rows in dataframe
+* Aggregation methods 
+    * Dataframe.describe(); // return insights in dataframe like mean, count,  min, max
+* Viewing data 
+    * Dataframe.print(); // print the scheme of dataframe
+* Group by one or two columns 
+
+Engineering Infrastructure
+==========================
+
+* Automatic build system: make
+* Version control: git
+* Testing framework: pytest
+
+Schedule
+========
+
+* Planning phase (6 weeks including 2/21, 2/28, 3/7, 3/14, 3/21, 3/28):
+* Week 1 (4/4): 
+    * Implement the main class, and some of the method.
+   	* Build automatic python unit test to test the correctness of output of C++ module.
+* Week 2 (4/11):
+    * Build the rest of method. 
+	* Analyze the memory behavior in runtime, refined the previous design.
```

Comment:
Memory analysis is non-trivial.  It may help you to think about what tools you want to use for it.